dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -1,46 +1,142 @@
 package com.dat3m.dartagnan.program.processing;
 
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionFactory;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
 import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Register;
 import com.dat3m.dartagnan.program.event.EventFactory;
 import com.dat3m.dartagnan.program.event.core.CondJump;
 import com.dat3m.dartagnan.program.event.core.Label;
-
+import com.dat3m.dartagnan.program.event.core.Local;
+import java.math.BigInteger;
 import java.util.List;
 import java.util.Map;
 
 /*
-    This pass transforms loops to have a single backjump using forward jumping.
-    Given a loop of the form
+    This pass normalizes loops to have a single unconditional backjump and a single entry point.
+    It achieves this via two transformations.
+
+    (1) Given a loop of the form
 
-    L:
-    ...
-    if X goto L
-    ...
-    if Y goto L
-    More Code
+        entry:
+        ...
+        goto C
+        ...
+        L:
+        ...
+        C:
+        ...
+        goto L
 
     it transforms it to
 
-    L:
-    ...
-    if X goto __repeatLoop_L
-    ...
-    if Y goto __repeatLoop_L
-    goto __breakLoop_L
-    __repeatLoop_L:
-    goto L
-    __breakLoop_L
-    More Code
-    ...
+        entry:
+        __jumpedTo_L_From <- 0
+        ...
+        __jumpedTo_L_From <- 1
+        goto L
+        ...
+        L:
+        if __jumpedTo_L_From == 1 goto C
+        ...
+        C:
+        ...
+        __jumpedTo_L_From <- 0
+        goto L
+
+    (2) Given a loop of the form
+
+        L:
+        ...
+        if X goto L
+        ...
+        if Y goto L
+        More Code
+
+    it transforms it to
+
+        L:
+        ...
+        if X goto __repeatLoop_L
+        ...
+        if Y goto __repeatLoop_L
+        goto __breakLoop_L
+        __repeatLoop_L:
+        goto L
+        __breakLoop_L
+        More Code
+        ...
 */
 public class NormalizeLoops implements FunctionProcessor {
 
+    private final TypeFactory types = TypeFactory.getInstance();
+    private final ExpressionFactory expressions = ExpressionFactory.getInstance();
+
     public static NormalizeLoops newInstance() {
         return new NormalizeLoops();
     }
 
     @Override
     public void run(Function function) {
```

Comment:
You should extract two helper methods from this: `guaranteeUniqueEntryPoint` and `guaranteeUniqueBackjump`.